File: repls/repl3.js
// Stateful modules

// Node.js style module system
function require(name) {
  return require.cache[name];
}
require.cache = {};
function define(name, body) {
  var module = {}, exports = {};
  module.exports = exports;
  body(module, exports);
  require.cache[name] = module.exports;
}

// And some modules

define('emoji', function(module, exports) {
  module.exports = {
    'horse': '🐴',
    'monkey': '🐵',
    'bear': '🐻'
  };
});

define('voting', function(module, exports) {
  var votes = exports.votes = {};
  exports.vote = function vote(name) {
    votes[name] = (votes[name] || 0) + 1;
  };
});

define('display', function(module, exports) {
  var emoji = require('emoji');

  function repeat(n, char) {
    if (!n) return '';
    return new Array(n + 1).join(char); // oops
  }

  module.exports = function display(votes) {
    Object.keys(emoji).forEach(function(name) {
      console.log(
        "%s\t%s", name,
        repeat(votes[name], emoji[name])
      );
    });
  };
});

// Using modules

var voting = require('voting');
var display = require('display');

voting.vote('horse');
voting.vote('horse');
voting.vote('horse');
voting.vote('monkey');
voting.vote('monkey');
voting.vote('bear');

display(voting.votes);

File: repls/repl3.clj
; Stateful modules in clojure

(ns emoji)
(def items {:horse "🐴",
            :monkey "🐵",
            :bear "🐻"
            :rabbit "r"})

(ns voting)
(defonce votes (atom {}))
(defn vote [name]
  (swap! votes update-in [name] (fnil inc 0)))

(ns display
  (:require emoji clojure.string))
(defn- repeating [n char]
  (clojure.string/join (repeat n char)))
(defn display [votes]
  (reduce-kv
    (fn [_ k v]
      (println (name k) "\t"
               (repeating v (k emoji/items))))
    nil votes))

(ns user
  (:require voting display))

(voting/vote :horse)
(voting/vote :horse)
(voting/vote :horse)
(voting/vote :monkey)
(voting/vote :monkey)
(voting/vote :bear)

(display/display @voting/votes)
